[Feature] Generic secondary resources

InDebt opened the pull request:
## Generic Secondary resources

This PR will enable generic secondary resources that can be configured for each individual Sim UI to reflect their respective resource. 

Right now it can be used like this for example (note. the extra variable for ShadowOrb is just for taste and more expressivnes in the spell later)
in `shadow.go`
```go
spriest.RegisterSecondaryResourceBar(core.SecondaryResourceConfig{
    Type: core.ShadowOrbs,
    Max:  3,
})

spriest.ShadowOrbs = spriest.SecondaryResourceBar
```

in `mind_blast.go`
```go
shadow.ShadowOrbs.Gain(1, spell.ActionID, sim)
```

With a simple configuration like
```js
secondaryResource: {
    color: "#b8a8f0",
    icon: "https://wow.zamimg.com/images/wow/icons/large/spell_priest_shadoworbs.jpg",
    name: "Shadow Orb",
}
```

The UI will display the generic resource accordingly.

![grafik](https://github.com/user-attachments/assets/e2b8e08d-0226-433c-8774-154cb7da6b81)
![grafik](https://github.com/user-attachments/assets/1c407a44-6b41-4590-8bb0-3eb41ce5d2b0)

InDebt commented:
Looking into this - in theory we can add a new `SecondaryResourceCost` that would check for a certain secondary resource.
This resource has to be spent however, so spells that consume 'all' or 'UpTo' (i.E. 3 Holy Power) would either need a rework of the `ApplyEffect` method to supply the amount of resource used, or for those spells it needs to be done through the `ExtraCondition` and resources are spend in the `ApplyEffect` method.

This would be rather intransparent I feel. We can define a cost - but may only do so if the cost can not be spend before the ApplyEffect method. In my opinion it's easier and more transparent to simply define a `ExtraCondition` for all spells using a secondary resoruce and handle the appropiate spending logic within the spell.

Thoughts?

InDebt commented on ui/core/proto_utils/.action_id.ts.swp:
That is a good question :D

NerdEgghead commented on sim/core/resource_bar.go:
```diff
@@ -0,0 +1,143 @@
+	metrics := bar.GetMetric(action)
+	metrics.AddEvent(baseAmount, amount)
+	if sim.Log != nil {
+		bar.unit.Log(
+			sim,
+			"Gained %0.0f generic resource from %s (%0.0f --> %0.0f) of %0.0f total.",
```
If you implement an additional method
```
func (resourceType SecondaryResourceType) String() string {
```
that maps the enum values to strings, then you can replace "generic resource" here with a "%s" on `bar.config.Type` to produce a customized log statement.

InDebt replied:
If I log as specific resource I'd have to rewrite https://github.com/wowsims/mop/pull/54/files#diff-b2999a368646afb09ca8f129e450fe7c7342fd31b2bf54ba071d62e7d79538d1R824

This magic regex (propably as old as time) get's the resouce name and maps it onto the actual TS Type.
I can rewrite it to not match any resource at all but any text between `%0.0f` and `from` by using `(\S.+?\S)` I guess and then map any resource not mapped to generic instead of none here 

```js
export function stringToResourceType(str: string): ResourceType {
	for (const [key, val] of resourceNames) {
		if (val.toLowerCase() == str.toLowerCase()) {
			return key;
		}
	}
	return ResourceType.ResourceTypeNone;
}
```

InDebt replied:
Okay - done and mapped within the logs now as well

NerdEgghead commented on sim/core/resource_bar.go:
```diff
@@ -0,0 +1,143 @@
+
+// Spend implements SecondaryResourceBar.
+func (bar *_SecondaryResourceBar) Spend(amount float64, action ActionID, sim *Simulation) {
+	if amount > bar.value {
+		panic("Trying to spend more resource than is available.")
+	}
```
Add an additional panic if `amount  < 0`.

NerdEgghead commented on sim/core/resource_bar.go:
```diff
@@ -0,0 +1,143 @@
+
+	return metric
+}
+
+func (unit *Unit) NewSecondaryResourceBar(config SecondaryResourceConfig) SecondaryResourceBar {
```
I would recommend merging this with `RegisterSecondaryResourceBar()` so that there is only one public access method containing all the validation checks.

NerdEgghead reviewed (changes requested):
I would also recommend ripping out the custom Holy Power resource implementation and replacing it with the new general framework as part of this PR. Now that we'll have the general version, we shouldn't keep any custom resource bars in the code anymore unless there's a good reason for it.

InDebt commented on sim/core/resource_bar.go:
```diff
@@ -0,0 +1,143 @@
+	DemonicFury   SecondaryResourceType = 104315
+)
+
+type SecondaryResourceBar interface {
+	CanSpend(limit float64) bool                                       // Check whether the current resource is available or not
+	Spend(amount float64, action ActionID, sim *Simulation)            // Spend the specified amount of resource
+	SpendUpTo(limit float64, action ActionID, sim *Simulation) float64 // Spends as much resource as possible up to the speciefied limit; Returns the amount of resource spent
+	Gain(amount float64, action ActionID, sim *Simulation)             // Gain the amount specified from the action
+	Reset(sim *Simulation)                                             // Resets the current resource bar
+	Value() float64                                                    // Returns the current amount of resource
+}
+
+type SecondaryResourceConfig struct {
+	Type    SecondaryResourceType // The type of resource the bar tracks
+	Max     float64               // The maximum amount the bar tracks
+	Default float64               // The default value this bar should be initialized with
+}
```
https://wago.tools/db2/PowerType?build=5.5.0.60548&filter%5BRegenCombat%5D=exact%3A0&page=1

Kind of - we have Soulshards and Embers where we would need to do some math like blizz does as they are capped at 4 but have a shift of 100 / 10 respectively. Similar to rage. Shards i.E. regen 0.05 Shards per Second and I'm not too sure about partial refunds of things - so I thought similar to rage we can just use float here as well.

NerdEgghead replied:
Is a fractional Soul Shard useful for anything though, or does a regen rate of 0.05 Shards per second really just mean 1 integer Shard every 20 seconds? Rage needs to be a float because of the refund mechanic: an unsuccessful single target cast with a Rage cost will refund 80% of the cost including the decimal component, so the full float needs to be tracked. Do Soul Shards have a fractional refund mechanic like this as well?

InDebt replied:
I have looked through some different documentation and moved the type to int32 - we can always change it later to float if needed. The shard regen is only out fight and can sureley be moddeled differently during the prepull if needed.

InDebt commented on sim/core/resource_bar.go:
```diff
@@ -0,0 +1,143 @@
+	Max     float64               // The maximum amount the bar tracks
+	Default float64               // The default value this bar should be initialized with
+}
+
+type _SecondaryResourceBar struct {
```
Yeah I'll do that. Usualy I would have implemented a `proctected` constructor with the factory method in place because escpecially when someone rarely uses it, it's easiy to just instantiate the struct and bypass the validation or setup in `Unit.go`. So this was the most intuitive way I could think of in go 😁 

I'll make it public though and leave a code comment so someone sees it hopefully if they were to directly instantiate the `DefaultSecondaryResourceBarImpl`

InDebt commented:
Cleaned some parts up based on feedback here. Also replaced the holy power implementation to be based on the SecondaryResourceBar. Due to Divine Purpose I actually just implemented a light weight `HolyPowerBar` that implements the Divine Purpose logic. 

After touching the Holy Power implementation though, I guess we also need to provider proper APL support.

InDebt commented:
With the Help of @1337LutZ APL support has been integrated now as well that dynamically resolves the resource name for the APL action. 

![grafik](https://github.com/user-attachments/assets/70c819f1-ca81-4028-93af-5b2b7847e878)

NerdEgghead commented on sim/paladin/holy_power_bar.go:
```diff
@@ -0,0 +1,60 @@
+
+import "github.com/wowsims/mop/sim/core"
+
+type HolyPowerBar struct {
+	resourceBar core.SecondaryResourceBar
+	paladin     *Paladin
+}
+
+// RegisterOnGain implements core.SecondaryResourceBar.
+func (h HolyPowerBar) RegisterOnGain(callback core.OnGainCallback) {
+	h.resourceBar.RegisterOnGain(callback)
+}
+
+// RegisterOnSpend implements core.SecondaryResourceBar.
+func (h HolyPowerBar) RegisterOnSpend(callback core.OnSpendCallback) {
+	h.resourceBar.RegisterOnSpend(callback)
+}
+
+// CanSpend implements core.SecondaryResourceBar.
+func (h HolyPowerBar) CanSpend(limit int32) bool {
+	return h.Value() >= limit
+}
+
+// Gain implements core.SecondaryResourceBar.
+func (h HolyPowerBar) Gain(amount int32, action core.ActionID, sim *core.Simulation) {
+	h.resourceBar.Gain(amount, action, sim)
+}
+
+// Reset implements core.SecondaryResourceBar.
+func (h HolyPowerBar) Reset(sim *core.Simulation) {
+	h.resourceBar.Reset(sim)
+}
```
You can delete all this shared boilerplate if you directly embed `DefaultSecondaryResourceBarImpl` in the struct:
```
type HolyPowerBar struct {
	*core.DefaultSecondaryResourceBarImpl
	paladin *Paladin
}
```
This guarantees that `HolyPowerBar` implements `core.SecondaryResourceBar`, so you only need to supply your own implementations of the few interface methods that differ from the default (`Spend()`, `SpendUpTo()`, and `Value()`).

NerdEgghead commented on sim/core/resource_bar.go:
```diff
@@ -0,0 +1,183 @@
+	}
+}
+
+func (unit *Unit) RegisterSecondaryResourceBar(config SecondaryResourceConfig) SecondaryResourceBar {
+	if config.Type <= 0 {
+		panic("Invalid SecondaryResourceType given.")
+	}
+
+	if config.Max <= 0 {
+		panic("Invalid maximum resource value given.")
+	}
+
+	if config.Default < 0 || config.Default > config.Max {
+		panic("Invalid default value given for resource bar")
+	}
+
+	if unit.SecondaryResourceBar != nil {
+		panic("A secondary resource bar has already been registered.")
+	}
+
+	unit.SecondaryResourceBar = &DefaultSecondaryResourceBarImpl{
+		config:  config,
+		unit:    unit,
+		metrics: make(map[ActionID]*ResourceMetrics),
+		onGain:  []OnGainCallback{},
+		onSpend: []OnSpendCallback{},
+	}
+
+	return unit.SecondaryResourceBar
+}
```
Now that I see the Paladin implementation, I realize that we do need to break this out into separate public access methods:
```
func (unit *Unit) NewDefaultSecondaryResourceBar(config SecondaryResourceConfig) *DefaultSecondaryResourceBarImpl {
```
and
```
func (unit *Unit) RegisterSecondaryResourceBar(impl SecondaryResourceBar) SecondaryResourceBar {
```
The two could be chained together with a third helper for specs that don't need any custom implementations:
```
func (unit *Unit) RegisterDefaultSecondaryResourceBar(config SecondaryResourceConfig) SecondaryResourceBar {
```

NerdEgghead commented on sim/paladin/paladin.go:
```diff
@@ -302,7 +300,15 @@ func NewPaladin(character *core.Character, talentsStr string, options *proto.Pal
 
 	paladin.EnableManaBar()
-	paladin.initializeHolyPowerBar()
+	paladin.HolyPower = HolyPowerBar{
+		resourceBar: paladin.RegisterSecondaryResourceBar(core.SecondaryResourceConfig{
+			Type:    proto.SecondaryResourceType_SecondaryResourceTypeHolyPower,
+			Max:     3,
+			Default: paladin.StartingHolyPower,
+		}),
+		paladin: paladin,
+	}
+	paladin.Unit.SecondaryResourceBar = paladin.HolyPower
```
With my previous suggestions, then this code would be
```
paladin.HolyPower = &HolyPowerBar{
    DefaultSecondaryResourceBarImpl: paladin.NewDefaultSecondaryResourceBar(core.SecondaryResourceConfig{
			Type:    proto.SecondaryResourceType_SecondaryResourceTypeHolyPower,
			Max:     3,
			Default: paladin.StartingHolyPower,
    }),

    paladin: paladin,
}

paladin.RegisterSecondaryResourceBar(paladin.HolyPower)
```

NerdEgghead reviewed (changes requested):
Getting closer!

InDebt commented on proto/spell.proto:
```diff
@@ -40,6 +40,6 @@ enum ResourceType {
 	ResourceTypeLunarEnergy = 13;
-	ResourceTypeHolyPower = 14;
-	ResourceTypeChi = 15;
+	ResourceTypeChi = 14;
+	ResourceTypeGenericResource = 15;
```
I'll have to request Hiller here I guess. Monk changes are already on master and they've somewhat already implemented a lot of chi stuff in the energy bar as comobpoint metric. I'm not to sure exactly if chi exactly behaves like combopoints.

@hillerstorm can you check if we can use the secondary resource bar for chi or do we need the combopoint specific logic for Monk?

NerdEgghead replied:
If Monk wants to use the Energy bar implementation for Chi that's fine too, but in that case we should remove `SecondaryResourceTypeChi` from `SecondaryResourceType` if we're modeling Chi as a primary resource instead.

hillerstorm commented:
> Cleaned some parts up based on feedback here. Also replaced the holy power implementation to be based on the SecondaryResourceBar. Due to Divine Purpose I actually just implemented a light weight `HolyPowerBar` that implements the Divine Purpose logic. 
> 
> 
> 
> After touching the Holy Power implementation though, I guess we also need to provider proper APL support.

In my local pala branch I've actually cleaned up the DP proc logic a bit by having the holy power bar have a pub/sub for resource gain/spent, I can just do that on top of these changes later 🙂 seems like a useful thing

hillerstorm commented:
> > Cleaned some parts up based on feedback here. Also replaced the holy power implementation to be based on the SecondaryResourceBar. Due to Divine Purpose I actually just implemented a light weight `HolyPowerBar` that implements the Divine Purpose logic. 
> > 
> > 
> > 
> > After touching the Holy Power implementation though, I guess we also need to provider proper APL support.
> 
> In my local pala branch I've actually cleaned up the DP proc logic a bit by having the holy power bar have a pub/sub for resource gain/spent, I can just do that on top of these changes later 🙂 seems like a useful thing 

And now I see you also added the pub/sub 😅 nice!

InDebt commented:
Alright - Chi is gone

NerdEgghead commented on sim/core/energy.go:
```diff
@@ -285,7 +285,7 @@ func newEnergyCost(spell *Spell, options EnergyCostOptions, energyBar *energyBar
 			RefundMetrics:     options.RefundMetrics,
 			ResourceMetrics:   spell.Unit.NewEnergyMetrics(spell.ActionID),
-			ComboPointMetrics: Ternary(energyBar.ownerClass == proto.Class_ClassMonk, spell.Unit.NewChiMetrics(spell.ActionID), spell.Unit.NewComboPointMetrics(spell.ActionID)),
+			ComboPointMetrics: spell.Unit.NewComboPointMetrics(spell.ActionID),
```
Delete the `NewChiMetrics()` method as well if it is no longer needed.

NerdEgghead reviewed (approved):
Looks great, just added one additional comment.
